[SPARK-6518][MLlib][Example][DOC] Add example code and user guide for bisecting k-means

yu-iskw opened the pull request:
This PR includes only an example code in order to finish it quickly.
I'll send another PR for the docs soon.

yu-iskw committed:
[SPARK-6518][MLlib] Add example code and user guide for bisecting k-means

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46655/
Test FAILed.

yu-iskw commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #46657 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46657/consoleFull)** for PR 9952 at commit [`7c8005d`](https://github.com/apache/spark/commit/7c8005d51283c818c6dde4737ae89571ad0d417d).

SparkQA commented:
**[Test build #46657 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46657/consoleFull)** for PR 9952 at commit [`7c8005d`](https://github.com/apache/spark/commit/7c8005d51283c818c6dde4737ae89571ad0d417d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `  case class Params(`\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46657/
Test PASSed.

yu-iskw commented:
@jkbradley could you review this PR?

yu-iskw committed:
[SPARK-6518][MLlib][DOC] Add example code and user guide for bisecting k-means

jkbradley commented:
When you modify this and [https://github.com/apache/spark/pull/9968] to use the ```include_example``` functionality, it may make sense to change this example code to use fixed parameters, rather than making it a command-line tool.

yu-iskw commented:
All right. I'll fix it.

yu-iskw committed:
Use the fixed paramters

SparkQA commented:
**[Test build #46790 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46790/consoleFull)** for PR 9952 at commit [`86f6085`](https://github.com/apache/spark/commit/86f608546bfac909eb4e99bf49816596767feac8).

SparkQA commented:
**[Test build #46790 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46790/consoleFull)** for PR 9952 at commit [`86f6085`](https://github.com/apache/spark/commit/86f608546bfac909eb4e99bf49816596767feac8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46790/
Test PASSed.

yu-iskw commented:
@jkbradley could you review it?

yu-iskw committed:
Delete the sample data

yu-iskw committed:
Replace the sample code with `include_example`

yu-iskw committed:
Add $example on$ and $example off$ to the example code

yu-iskw commented:
I moved the documentation to this PR from https://github.com/apache/spark/pull/9968. Because the PR depends on this PR.

I modified the docs for `BisectingKMeans` with `include_example`.

SparkQA commented:
**[Test build #47095 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47095/consoleFull)** for PR 9952 at commit [`2f4c2a5`](https://github.com/apache/spark/commit/2f4c2a5599a4d9383773ba5cd47f0343106d713b).

SparkQA commented:
**[Test build #47095 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47095/consoleFull)** for PR 9952 at commit [`2f4c2a5`](https://github.com/apache/spark/commit/2f4c2a5599a4d9383773ba5cd47f0343106d713b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47095/
Test PASSed.

yu-iskw commented:
@jkbradley ping

jkbradley commented:
Thanks!  I'll review today

jkbradley commented:
I'll go ahead and review this, but if you have time to add a Java example, that would be awesome--thanks!

jkbradley commented on docs/mllib-clustering.md:
```diff
@@ -718,6 +718,34 @@ sameModel = LDAModel.load(sc, "myModelPath")
+
+Bisecting k-means is a kind of [hierarchical clustering](https://en.wikipedia.org/wiki/Hierarchical_clustering).
+Hierarchical clustering is one of the most commonly used  method of cluster analysis which seeks to build a hierarchy of clusters.
+Strategies for hierarchical clustering generally fall into two types:
```
Before this line (which is getting into details), it would be good to give a high-level description of the use of bisecting k-means.  E.g., "Bisecting K-means can often be much faster than regular K-means, but it will generally produce a different clustering."

jkbradley commented on docs/mllib-clustering.md:
```diff
@@ -718,6 +718,34 @@ sameModel = LDAModel.load(sc, "myModelPath")
+- Divisive: This is a "top down" approach: all observations start in one cluster, and splits are performed recursively as one moves down the hierarchy.
+
+Bisecting k-means algorithm is a kind of divisive algorithms.
+Because it is too difficult to implement a agglomerative algorithm as a distributed algorithm on Spark.
```
Remove this line please; it may not always be the case.

jkbradley commented on docs/mllib-clustering.md:
```diff
@@ -718,6 +718,34 @@ sameModel = LDAModel.load(sc, "myModelPath")
+Because it is too difficult to implement a agglomerative algorithm as a distributed algorithm on Spark.
+The implementation in MLlib has the following parameters:
+
+* *k* the desired number of leaf clusters (default: 4). The actual number could be smaller if there are no divisible leaf clusters.
```
put colon between parameter name and the description

jkbradley commented on examples/src/main/scala/org/apache/spark/examples/mllib/BisectingKMeansExample.scala:
```diff
@@ -0,0 +1,58 @@
+package org.apache.spark.examples.mllib
+
+// scalastyle:off println
+import org.apache.spark.mllib.clustering.BisectingKMeans
```
Can you please use example on/off here too to include the relevant imports?  You can exclude SparkConf, SparkContext.

jkbradley commented:
That should be it.

jkbradley commented:
Also, can you please add the "[DOC]" tag to the PR title?  T'hanks!

yu-iskw commented:
@jkbradley sure, done.

jkbradley commented:
Thanks, I'll check back for updates for the few remaining comments.

yu-iskw commented:
ping @jkbradley

jkbradley commented:
@yu-iskw There are several comments of mine (above) which have not yet been addressed.  Could you please update this PR based on those comments?  I'll then take another look.

yu-iskw committed:
Modify the documentation and add a Java example code

yu-iskw commented:
Oh, I'm terribly sorry about that. Pushing the update was failed.... 
- Modified what you pointed out
- Add a Java example and its doc

SparkQA commented:
**[Test build #47639 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47639/consoleFull)** for PR 9952 at commit [`7c36c1d`](https://github.com/apache/spark/commit/7c36c1dc34074712a19d227ba69cfd64178a82f1).

SparkQA commented:
**[Test build #47639 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47639/consoleFull)** for PR 9952 at commit [`7c36c1d`](https://github.com/apache/spark/commit/7c36c1dc34074712a19d227ba69cfd64178a82f1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `public class JavaBisectingKMeansExample `\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47639/
Test PASSed.

jkbradley commented on examples/src/main/java/org/apache/spark/examples/mllib/JavaBisectingKMeansExample.java:
```diff
@@ -0,0 +1,68 @@
+
+import org.apache.spark.SparkConf;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.JavaSparkContext;
```
I would exclude JavaSparkContext and SparkConf.

jkbradley commented on examples/src/main/java/org/apache/spark/examples/mllib/JavaBisectingKMeansExample.java:
```diff
@@ -0,0 +1,68 @@
+
+    // $example on$
+    ArrayList<Vector> localData = Lists.newArrayList(
+        Vectors.dense(0.1, 0.1),   Vectors.dense(0.3, 0.3),
```
Indent with 2 spaces in Java please

jkbradley commented on examples/src/main/java/org/apache/spark/examples/mllib/JavaBisectingKMeansExample.java:
```diff
@@ -0,0 +1,68 @@
+    JavaRDD<Vector> data = sc.parallelize(localData, 2);
+
+    BisectingKMeans bkm = new BisectingKMeans()
+        .setK(4);
```
fix indentation here too

jkbradley commented:
No problem.  I just added some comments and am generating the docs now.

jkbradley commented:
The generated docs look good, so just those few comments.

yu-iskw committed:
Modify the wrong indentations

yu-iskw committed:
Exclude the import statments from SparkConf and JavaSparkContext in the docs

yu-iskw commented:
@jkbradley thanks for the review. I modified the wrong indentations and excluded the import statements for `SparkConf` and `JavaSparkContext`.

SparkQA commented:
**[Test build #47780 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47780/consoleFull)** for PR 9952 at commit [`3066696`](https://github.com/apache/spark/commit/3066696a16709f58fe5cc179274b0a7ba277481a).

SparkQA commented:
**[Test build #47780 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47780/consoleFull)** for PR 9952 at commit [`3066696`](https://github.com/apache/spark/commit/3066696a16709f58fe5cc179274b0a7ba277481a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `public class JavaBisectingKMeansExample `\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47780/
Test PASSed.

jkbradley commented:
LGTM, merging with master and branch-1.6
Thanks!

yu-iskw commented:
Thank you for merging it!
